### ISSUE TITLE
[HIP][O2][Multigrid test fails validating (1 - P P^\dagger) v_k test for k >=2 on MG level 1]

### ISSUE THREAD

**Issue by bjoo** [lattice/quda#1134]
When generating MG Projectors for multigrid the (1 - P P^\dagger) v_k test fails on MG level 1, when using -O2 in the build and tuning is enabled (ROCm-4.1.0 RC2, QUDA config=DEVEL)

Turning off tuning by setting env var `QUDA_ENABLE_TUNING=0` circumvents problem 

-O1 results in a `hipOutOfMemory` error  see #1131

**Comment by bjoo** [comment]
MG script was:

```
export META="--unit-gauge 0 --dim 16 16 16 16 --Lsdim 1 --gridsize 1 1 1 1 --solve-type direct-pc --dslash-type clover --mass 0 --clover-coeff 0.01 --compute-clover 1 --anisotropy 1 --inv-multigrid true"
export SOLVER_PARAM="--prec single --prec-sloppy single --prec-precondition single --tol 5e-7"
export MG_PARAM="--mg-levels 3 --mg-generate-nullspace true"
export LEVEL0_PARAM="--mg-nu-pre 0 0 --mg-nu-post 0 8 --mg-block-size 0 4 4 4 4 --mg-nvec 0 24 --mg-setup-inv 0 cg --mg-smoother 0 ca-gcr"
export LEVEL1_PARAM="--mg-nu-pre 1 0 --mg-nu-post 1 8 --mg-block-size 1 2 2 2 2 --mg-nvec 1 32 --mg-setup-inv 1 cg --mg-smoother 1 ca-gcr"
export LEVEL2_PARAM="--mg-coarse-solver 2 ca-gcr --mg-coarse-solver-maxiter 2 8"
export DIAG_PARAMS="--verbosity verbose --mg-verbosity 0 verbose --mg-verbosity 1 verbose"
./tests/invert_test $META $SOLVER_PARAM $MG_PARAM $LEVEL0_PARAM $LEVEL1_PARAM $LEVEL2_PARAM $DIAG_PARAMS
```

run from the top build directory

**Comment by mathiaswagner** [comment]
@bjoo Can you please apply the 'Target_HIP' label for HIP bugs?

**Comment by bjoo** [comment]
Sure thing Mathias, we also grouped them under a HIP_Milestone. If you don’t want HIP issues tracked here let us know.

Best,
B

--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: Mathias Wagner ***@***.***>
Reply-To: lattice/quda ***@***.***>
Date: Wednesday, April 7, 2021 at 3:39 PM
To: lattice/quda ***@***.***>
Cc: Balint Joo ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [lattice/quda] [HIP][O2][Multigrid test fails validating (1 - P P^dagger) v_k test for k >=2 on MG level 1] (#1134)
Resent-From: ***@***.***>
Resent-Date: Wednesday, April 7, 2021 at 3:39 PM


@bjoo<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_bjoo&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=Dml58MAvYoTf6ArtBiqda5jBBQbbbtrR6nC_3McXoTY&s=2a7Mv23IMjgbhy2CHjLfANk0OU3lvturt3x1HsSReKk&e=> Can you please apply the 'Target_HIP' label for HIP bugs?

**Comment by mathiaswagner** [comment]
They can stay here. All good. The label just helps to quickly identify them on the issues page (it is more prominent then the milestone) or filter them there.

**Comment by bjoo** [comment]
Issue was fixed in commit f03dd39 by making the minimum block size equal to device::warp_size()
